File: app/services/file_manager.py
"""File manager service for app operations."""

import asyncio
import re
from datetime import datetime
from pathlib import Path
from typing import Any

import yaml

from app.config import get_settings
from app.services.version_control import VersionControl


class AppInfo:
    """Information about an AppDaemon app."""

    def __init__(
        self,
        name: str,
        class_name: str,
        description: str,
        has_python: bool,
        has_yaml: bool,
        last_modified: datetime,
        version_count: int,
    ):
        self.name = name
        self.class_name = class_name
        self.description = description
        self.has_python = has_python
        self.has_yaml = has_yaml
        self.last_modified = last_modified
        self.version_count = version_count


class FileManagerError(Exception):
    """Base exception for file manager errors."""

    pass


class InvalidAppNameError(FileManagerError):
    """Raised when app name is invalid."""

    pass


class AppNotFoundError(FileManagerError):
    """Raised when app is not found."""

    pass


class PathTraversalError(FileManagerError):
    """Raised when path traversal is detected."""

    pass


class FileManager:
    """Manages AppDaemon app files."""

    def __init__(self, base_path: Path | None = None):
        """Initialize file manager.

        Args:
            base_path: Base path for apps directory. Defaults to config setting.
        """
        self.base_path = base_path or get_settings().apps_path
        self.version_control = VersionControl(self.base_path)
        self._lock = asyncio.Lock()

    def _validate_app_name(self, name: str) -> None:
        """Validate app name for security.

        Args:
            name: App name to validate.

        Raises:
            InvalidAppNameError: If name is invalid.
        """
        if not name:
            raise InvalidAppNameError("App name cannot be empty")

        if not re.match(r"^[a-zA-Z][a-zA-Z0-9_]*$", name):
            raise InvalidAppNameError(
                "App name must start with a letter and contain only letters, numbers, and underscores"
            )

        if name.startswith(".") or name.startswith("_"):
            raise InvalidAppNameError("App name cannot start with . or _")

    def _get_app_path(self, name: str) -> Path:
        """Get validated app directory path.

        Args:
            name: App name.

        Returns:
            Path to app directory.

        Raises:
            PathTraversalError: If path traversal is detected.
        """
        self._validate_app_name(name)
        app_path = self.base_path / name

        # Prevent path traversal
        try:
            app_path.resolve().relative_to(self.base_path.resolve())
        except ValueError as e:
            raise PathTraversalError(f"Invalid path: {name}") from e

        return app_path

    def _get_python_path(self, name: str) -> Path:
        """Get Python file path for an app."""
        return self._get_app_path(name) / f"{name}.py"

    def _get_yaml_path(self, name: str) -> Path:
        """Get YAML file path for an app."""
        return self._get_app_path(name) / f"{name}.yaml"

    async def list_apps(self) -> list[AppInfo]:
        """List all apps in the apps directory.

        Returns:
            List of AppInfo objects.
        """
        apps: list[AppInfo] = []

        if not self.base_path.exists():
            return apps

        for item in self.base_path.iterdir():
            if item.is_dir() and not item.name.startswith("."):
                try:
                    app_info = await self._get_app_info(item.name)
                    apps.append(app_info)
                except (InvalidAppNameError, PathTraversalError):
                    # Skip invalid directories
                    continue

        # Sort by name
        apps.sort(key=lambda x: x.name)
        return apps

    async def _get_app_info(self, name: str) -> AppInfo:
        """Get information about a specific app.

        Args:
            name: App name.

        Returns:
            AppInfo object.
        """
        python_path = self._get_python_path(name)
        yaml_path = self._get_yaml_path(name)

        has_python = python_path.exists()
        has_yaml = yaml_path.exists()

        # Get last modified time
        last_modified = datetime.now()
        if has_python:
            stat = python_path.stat()
            last_modified = datetime.fromtimestamp(stat.st_mtime)
        elif has_yaml:
            stat = yaml_path.stat()
            last_modified = datetime.fromtimestamp(stat.st_mtime)

        # Parse class name and description from Python file
        class_name = name
        description = ""

        if has_python:
            try:
                content = await self.read_python(name)
                class_match = re.search(r"class\s+(\w+)", content)
                if class_match:
                    class_name = class_match.group(1)

                doc_match = re.search(r'"""(.+?)"""', content, re.DOTALL)
                if doc_match:
                    description = doc_match.group(1).strip().split("\n")[0]
            except Exception:
                pass

        # Get version count
        versions = await self.version_control.list_versions(name)
        version_count = len(versions)

        return AppInfo(
            name=name,
            class_name=class_name,
            description=description,
            has_python=has_python,
            has_yaml=has_yaml,
            last_modified=last_modified,
            version_count=version_count,
        )

    async def create_app(self, name: str, class_name: str, description: str = "") -> None:
        """Create a new app with template.

        Args:
            name: App name.
            class_name: Class name for the app.
            description: Optional description.

        Raises:
            InvalidAppNameError: If name is invalid.
            FileManagerError: If app already exists.
        """
        async with self._lock:
            self._validate_app_name(name)

            app_path = self._get_app_path(name)

            if app_path.exists():
                raise FileManagerError(f"App '{name}' already exists")

            # Create app directory
            app_path.mkdir(parents=True, exist_ok=False)

            # Create Python file from template
            python_content = self._render_template(name, class_name, description)
            python_path = self._get_python_path(name)

            loop = asyncio.get_event_loop()
            await loop.run_in_executor(
                None, lambda: python_path.write_text(python_content, encoding="utf-8")
            )

            # Create empty YAML file
            yaml_path = self._get_yaml_path(name)
            yaml_content = self._render_yaml_template(name, class_name)

            await loop.run_in_executor(
                None, lambda: yaml_path.write_text(yaml_content, encoding="utf-8")
            )

    def _render_template(self, name: str, class_name: str, description: str) -> str:
        """Render Python app template."""
        from jinja2 import Template

        template_path = Path(__file__).parent.parent / "templates" / "empty.py.j2"
        template_content = template_path.read_text(encoding="utf-8")

        template = Template(template_content)
        return template.render(
            name=name, class_name=class_name, description=description or f"AppDaemon app: {name}"
        )

    def _render_yaml_template(self, name: str, class_name: str) -> str:
        """Render YAML configuration template."""
        config = {name: {"module": name, "class": class_name}}
        return yaml.dump(config, default_flow_style=False, sort_keys=False)

    async def read_python(self, name: str) -> str:
        """Read Python file content.

        Args:
            name: App name.

            Returns:
            Python file content.

        Raises:
            AppNotFoundError: If app doesn't exist.
        """
        app_path = self._get_app_path(name)

        if not app_path.exists():
            raise AppNotFoundError(f"App '{name}' not found")

        python_path = self._get_python_path(name)

        if not python_path.exists():
            raise AppNotFoundError(f"Python file for app '{name}' not found")

        loop = asyncio.get_event_loop()
        return await loop.run_in_executor(None, lambda: python_path.read_text(encoding="utf-8"))

    async def write_python(self, name: str, content: str) -> None:
        """Write Python file content with automatic backup.

        Args:
            name: App name.
            content: Python code content.

        Raises:
            AppNotFoundError: If app doesn't exist.
        """
        async with self._lock:
            app_path = self._get_app_path(name)

            if not app_path.exists():
                raise AppNotFoundError(f"App '{name}' not found")

            python_path = self._get_python_path(name)

            # Create backup of existing file
            if python_path.exists():
                existing_content = await self.read_python(name)
                await self.version_control.create_version(name, f"{name}.py", existing_content)

            # Write new content
            loop = asyncio.get_event_loop()
            await loop.run_in_executor(
                None, lambda: python_path.write_text(content, encoding="utf-8")
            )

            # Cleanup old versions
            await self.version_control.cleanup_old_versions(name)

    async def read_yaml(self, name: str) -> dict:
        """Read YAML configuration.

        Args:
            name: App name.

        Returns:
            YAML configuration as dictionary.

        Raises:
            AppNotFoundError: If app doesn't exist.
        """
        app_path = self._get_app_path(name)

        if not app_path.exists():
            raise AppNotFoundError(f"App '{name}' not found")

        yaml_path = self._get_yaml_path(name)

        if not yaml_path.exists():
            return {}

        loop = asyncio.get_event_loop()
        content = await loop.run_in_executor(None, lambda: yaml_path.read_text(encoding="utf-8"))

        try:
            return yaml.safe_load(content) or {}
        except yaml.YAMLError as e:
            raise FileManagerError(f"Invalid YAML: {e}") from e

    async def write_yaml(self, name: str, config: dict) -> None:
        """Write YAML configuration with automatic backup.

        Args:
            name: App name.
            config: Configuration dictionary.

        Raises:
            AppNotFoundError: If app doesn't exist.
        """
        async with self._lock:
            app_path = self._get_app_path(name)

            if not app_path.exists():
                raise AppNotFoundError(f"App '{name}' not found")

            yaml_path = self._get_yaml_path(name)

            # Create backup of existing file
            if yaml_path.exists():
                try:
                    existing_config = await self.read_yaml(name)
                    existing_content = yaml.dump(
                        existing_config, default_flow_style=False, sort_keys=False
                    )
                    await self.version_control.create_version(
                        name, f"{name}.yaml", existing_content
                    )
                except FileManagerError:
                    # If read fails, still try to write
                    pass

            # Write new content
            content = yaml.dump(config, default_flow_style=False, sort_keys=False)

            loop = asyncio.get_event_loop()
            await loop.run_in_executor(
                None, lambda: yaml_path.write_text(content, encoding="utf-8")
            )

            # Cleanup old versions
            await self.version_control.cleanup_old_versions(name)

    async def delete_app(self, name: str) -> None:
        """Delete an app (moves to .trash).

        Args:
            name: App name.

        Raises:
            AppNotFoundError: If app doesn't exist.
        """
        async with self._lock:
            app_path = self._get_app_path(name)

            if not app_path.exists():
                raise AppNotFoundError(f"App '{name}' not found")

            # Move to trash instead of deleting
            trash_path = (
                self.base_path / ".trash" / f"{name}_{datetime.now().strftime('%Y%m%d_%H%M%S')}"
            )
            trash_path.parent.mkdir(parents=True, exist_ok=True)

            loop = asyncio.get_event_loop()
            await loop.run_in_executor(None, lambda: app_path.rename(trash_path))

    async def app_exists(self, name: str) -> bool:
        """Check if app exists.

        Args:
            name: App name.

        Returns:
            True if app exists, False otherwise.
        """
        try:
            app_path = self._get_app_path(name)
            return app_path.exists() and app_path.is_dir()
        except (InvalidAppNameError, PathTraversalError):
            return False

File: src/lib/file-manager.ts
import { promises as fs } from 'fs';
import path from 'path';
import { AppInfo, CreateAppData, FileContent } from '../types';
import { generatePythonTemplate, generateYamlTemplate } from './templates';

const APPS_DIR = '/config/apps';

export class AppNotFoundError extends Error {
  constructor(appName: string) {
    super(`App '${appName}' not found`);
    this.name = 'AppNotFoundError';
  }
}

export class InvalidAppNameError extends Error {
  constructor(name: string) {
    super(`Invalid app name: '${name}'. Use only alphanumeric characters and underscores.`);
    this.name = 'InvalidAppNameError';
  }
}

export class FileManagerError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'FileManagerError';
  }
}

function validateAppName(name: string): void {
  if (!name || !/^[a-zA-Z_][a-zA-Z0-9_]*$/.test(name)) {
    throw new InvalidAppNameError(name);
  }
}

export async function listApps(): Promise<AppInfo[]> {
  const apps: AppInfo[] = [];
  
  try {
    await fs.access(APPS_DIR);
  } catch {
    await fs.mkdir(APPS_DIR, { recursive: true });
    return apps;
  }
  
  const entries = await fs.readdir(APPS_DIR, { withFileTypes: true });
  
  for (const entry of entries) {
    if (entry.isDirectory()) {
      const appPath = path.join(APPS_DIR, entry.name);
      const pythonPath = path.join(appPath, `${entry.name}.py`);
      const yamlPath = path.join(appPath, `${entry.name}.yaml`);
      
      const [hasPython, hasYaml, stats] = await Promise.all([
        fs.access(pythonPath).then(() => true).catch(() => false),
        fs.access(yamlPath).then(() => true).catch(() => false),
        fs.stat(appPath)
      ]);
      
      // Get class name from Python file if exists
      let className = entry.name;
      if (hasPython) {
        try {
          const content = await fs.readFile(pythonPath, 'utf-8');
          const match = content.match(/class\s+(\w+)\s*\(/);
          if (match) {
            className = match[1];
          }
        } catch {
          // Use directory name
        }
      }
      
      apps.push({
        name: entry.name,
        class_name: className,
        description: '', // Could read from YAML or Python docstring
        has_python: hasPython,
        has_yaml: hasYaml,
        last_modified: stats.mtime.toISOString(),
        version_count: 0 // Will be populated by version control
      });
    }
  }
  
  return apps.sort((a, b) => a.name.localeCompare(b.name));
}

export async function createApp(data: CreateAppData): Promise<AppInfo> {
  validateAppName(data.name);
  
  const appDir = path.join(APPS_DIR, data.name);
  
  try {
    await fs.access(appDir);
    throw new FileManagerError(`App '${data.name}' already exists`);
  } catch (error) {
    if (error instanceof FileManagerError) throw error;
    // Directory doesn't exist, which is what we want
  }
  
  await fs.mkdir(appDir, { recursive: true });
  
  // Create initial Python file
  const pythonContent = generatePythonTemplate(data.name, data.class_name, data.description);
  await fs.writeFile(path.join(appDir, `${data.name}.py`), pythonContent);
  
  // Create initial YAML file
  const yamlContent = generateYamlTemplate(data.name, data.class_name);
  await fs.writeFile(path.join(appDir, `${data.name}.yaml`), yamlContent);
  
  return {
    name: data.name,
    class_name: data.class_name,
    description: data.description || '',
    has_python: true,
    has_yaml: true,
    last_modified: new Date().toISOString(),
    version_count: 0
  };
}

export async function deleteApp(name: string): Promise<void> {
  validateAppName(name);
  
  const appDir = path.join(APPS_DIR, name);
  
  try {
    await fs.access(appDir);
  } catch {
    throw new AppNotFoundError(name);
  }
  
  await fs.rm(appDir, { recursive: true });
}

export async function readPythonFile(appName: string): Promise<FileContent> {
  validateAppName(appName);
  
  const filePath = path.join(APPS_DIR, appName, `${appName}.py`);
  
  try {
    const [content, stats] = await Promise.all([
      fs.readFile(filePath, 'utf-8'),
      fs.stat(filePath)
    ]);
    
    return {
      content,
      last_modified: stats.mtime.toISOString()
    };
  } catch {
    throw new AppNotFoundError(appName);
  }
}

export async function writePythonFile(appName: string, content: string): Promise<void> {
  validateAppName(appName);
  
  const appDir = path.join(APPS_DIR, appName);
  const filePath = path.join(appDir, `${appName}.py`);
  
  try {
    await fs.access(appDir);
  } catch {
    throw new AppNotFoundError(appName);
  }
  
  await fs.writeFile(filePath, content);
}

export async function readYamlFile(appName: string): Promise<FileContent> {
  validateAppName(appName);
  
  const filePath = path.join(APPS_DIR, appName, `${appName}.yaml`);
  
  try {
    const [content, stats] = await Promise.all([
      fs.readFile(filePath, 'utf-8'),
      fs.stat(filePath)
    ]);
    
    return {
      content,
      last_modified: stats.mtime.toISOString()
    };
  } catch {
    throw new AppNotFoundError(appName);
  }
}

export async function writeYamlFile(appName: string, content: string): Promise<void> {
  validateAppName(appName);
  
  const appDir = path.join(APPS_DIR, appName);
  const filePath = path.join(appDir, `${appName}.yaml`);
  
  try {
    await fs.access(appDir);
  } catch {
    throw new AppNotFoundError(appName);
  }
  
  await fs.writeFile(filePath, content);
}

function parseYaml(content: string): unknown {
  // Simple YAML parser - returns first key's value
  const lines = content.split('\n');
  const result: Record<string, unknown> = {};
  let currentKey = '';
  
  for (const line of lines) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith('#')) continue;
    
    if (!line.startsWith(' ') && trimmed.endsWith(':')) {
      currentKey = trimmed.slice(0, -1);
      result[currentKey] = {};
    } else if (currentKey && line.startsWith('  ')) {
      const [key, ...valueParts] = trimmed.split(':');
      if (key && valueParts.length > 0) {
        (result[currentKey] as Record<string, string>)[key.trim()] = valueParts.join(':').trim();
      }
    }
  }
  
  return result;
}

function stringifyYaml(obj: Record<string, unknown>): string {
  // Simple YAML stringifier
  let result = '';
  
  for (const [key, value] of Object.entries(obj)) {
    result += `${key}:\n`;
    if (typeof value === 'object' && value !== null) {
      for (const [subKey, subValue] of Object.entries(value as Record<string, string>)) {
        result += `  ${subKey}: ${subValue}\n`;
      }
    }
  }
  
  return result;
}
